src/components/checkbox/checkbox.test.jsx

```diff
@@ -0,0 +1,233 @@
+      expect(ref.current.checked).toBe(true);
+    });
+  });
+
+  describe('value api', () => {
+    it('prioritizes the value provided', () => {
+      const ref = createRef();
+      render(<Checkbox {...requiredProps} value="i-am-good-bruh" ref={ref} />);
+      expect(ref.current.value).toBe('i-am-good-bruh');
+    });
+
+    it('sets the value to on if checked', () => {
+      const ref = createRef();
+      render(<Checkbox {...requiredProps} checked ref={ref} />);
+      expect(ref.current.value).toBe('on');
+    });
+
+    it('sets the value to off if not checked', () => {
+      const ref = createRef();
+      render(<Checkbox {...requiredProps} checked={false} ref={ref} />);
+      expect(ref.current.value).toBe('off');
+    });
+  });
+
+  describe('dirty api', () => {
+    it('is unset without any changes', () => {
+      const ref = createRef();
+      render(<Checkbox {...requiredProps} ref={ref} />);
+      expect(ref.current.dirty).toBe(false);
+    });
+
+    it('is set with any changes', () => {
+      const ref = createRef();
+      render(<Checkbox {...requiredProps} ref={ref} />);
+      userEvent.click(screen.getByRole('checkbox', { name }));
+      expect(ref.current.dirty).toBe(true);
+    });
+
+    it('is unset with changes back to initial state', () => {
+      const ref = createRef();
+      render(<Checkbox {...requiredProps} ref={ref} checked />);
+      userEvent.click(screen.getByRole('checkbox', { name }));
+      expect(ref.current.dirty).toBe(true);
+
+      userEvent.click(screen.getByRole('checkbox', { name }));
+      expect(ref.current.dirty).toBe(false);
+    });
+
+    it('is unset with changes and new value prop', () => {
+      const ref = createRef();
+      const { rerender } = render(
+        <Checkbox {...requiredProps} ref={ref} checked={true} />,
+      );
+      userEvent.click(screen.getByRole('checkbox', { name }));
+      expect(ref.current.dirty).toBe(true);
+
+      rerender(<Checkbox {...requiredProps} ref={ref} checked={false} />);
+      expect(ref.current.dirty).toBe(false);
+    });
+  });
+
+  describe('required api', () => {
+    it('can be set', () => {
+      render(<Checkbox {...requiredProps} required />);
+      expect(screen.getByRole('checkbox', { name })).toBeRequired();
+      expect(screen.getByRole('checkbox', { name })).toBeInvalid();
+    });
+
+    it('can be unset', () => {
+      render(<Checkbox {...requiredProps} required={false} />);
+      expect(screen.getByRole('checkbox', { name })).not.toBeRequired();
+      expect(screen.getByRole('checkbox', { name })).toBeValid();
+    });
+
+    it('is invalid on mount but does not have an error message', () => {
+      render(<Checkbox {...requiredProps} required />);
+      expect(screen.getByRole('checkbox', { name })).toBeInvalid();
+      expect(screen.getByRole('checkbox', { name })).toHaveDescription('');
+    });
+
+    it('is invalid after deselecting', () => {
```

Comment:
We should only be validating "on blur"

Reply:
5fa726efb5484c8688c48a615076de349af0d5c1